### ISSUE TITLE
Cleaner exception error

### ISSUE THREAD

**Issue by remondn-swk** [Slimmer-AI/mpire#16]
Currently if there is an exception, the stack trace is really convoluted :

> KeyError: '\n\nException occurred in Worker-0 with the following arguments:\nArg 0: 2\nTraceback (most recent call last):\n  File "C:\\Users\\Nicolas\\AppData\\Roaming\\Python\\Python39\\site-packages\\mpire\\worker.py", line 301, in _run_safely\n    results = func()\n  File "C:\\Users\\Nicolas\\AppData\\Roaming\\Python\\Python39\\site-packages\\mpire\\worker.py", line 255, in _func\n    results = func(args)\n  File "C:\\Users\\Nicolas\\AppData\\Roaming\\Python\\Python39\\site-packages\\mpire\\worker.py", line 394, in _helper_func_with_idx\n    return args[0], self._call_func(func, args[1])\n  File "C:\\Users\\Nicolas\\AppData\\Roaming\\Python\\Python39\\site-packages\\mpire\\worker.py", line 421, in _call_func\n    return func(args)\n  File "C:\\Users\\Nicolas\\Documents\\workspace\\poubelle\\lolilol.py", line 9, in get_state\n    return worker_state["x"] * x\nKeyError: \'x\'\n'

---

I believe it's from this part :

https://github.com/Slimmer-AI/mpire/blob/9c6332fb6cc0d1971880f35a1477ec3d92b5154c/mpire/pool.py#L583

---

To have a cleaner stack trace, we should instead use `raise <exception> from <catched_exception>`.

**Comment by sybrenjansen** [comment]
Thnx for your suggestion. 

I haven't used `raise from` before, so just wanted to check with you if this is the right way to use it. Unfortunately, traceback objects can't be pickled (which makes sense). Instead, I pass on the exception type and traceback string from one of the workers to the main process and use `raise err(traceback_str)` there once all the workers are shut down. I can change this to `raise err from err(traceback_str)`. This will still raise the correct error type, which can be caught, but it now splits the error message in two.

This is the result of the change:

Before:
```
Traceback (most recent call last):
  File "/home/sybren.local/repos/mpire/test_file.py", line 20, in <module>
    main()
  File "/home/sybren.local/repos/mpire/test_file.py", line 14, in main
    results = pool.map(square, range(1000), progress_bar=True)
  File "/home/sybren.local/repos/mpire/mpire/pool.py", line 258, in map
    results = self.map_unordered(func, ((args_idx, args) for args_idx, args in enumerate(iterable_of_args)),
  File "/home/sybren.local/repos/mpire/mpire/pool.py", line 314, in map_unordered
    return list(self.imap_unordered(func, iterable_of_args, iterable_len, max_tasks_active, chunk_size,
  File "/home/sybren.local/repos/mpire/mpire/pool.py", line 519, in imap_unordered
    self._handle_exception(progress_bar_handler)
  File "/home/sybren.local/repos/mpire/mpire/pool.py", line 583, in _handle_exception
    raise err(traceback_str)
ValueError: 

Exception occurred in Worker-0 with the following arguments:
Arg 0: 100
Traceback (most recent call last):
  File "/home/sybren.local/repos/mpire/mpire/worker.py", line 301, in _run_safely
    results = func()
  File "/home/sybren.local/repos/mpire/mpire/worker.py", line 255, in _func
    results = func(args)
  File "/home/sybren.local/repos/mpire/mpire/worker.py", line 394, in _helper_func_with_idx
    return args[0], self._call_func(func, args[1])
  File "/home/sybren.local/repos/mpire/mpire/worker.py", line 421, in _call_func
    return func(args)
  File "/home/sybren.local/repos/mpire/test_file.py", line 8, in square
    raise ValueError("ERROR")
ValueError: ERROR
```

After:
```
ValueError: 

Exception occurred in Worker-0 with the following arguments:
Arg 0: 100
Traceback (most recent call last):
  File "/home/sybren.local/repos/mpire/mpire/worker.py", line 304, in _run_safely
    results = func()
  File "/home/sybren.local/repos/mpire/mpire/worker.py", line 258, in _func
    results = func(args)
  File "/home/sybren.local/repos/mpire/mpire/worker.py", line 397, in _helper_func_with_idx
    return args[0], self._call_func(func, args[1])
  File "/home/sybren.local/repos/mpire/mpire/worker.py", line 424, in _call_func
    return func(args)
  File "/home/sybren.local/repos/mpire/test_file.py", line 8, in square
    raise ValueError("ERROR")
ValueError: ERROR


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/sybren.local/repos/mpire/test_file.py", line 24, in <module>
    main()
  File "/home/sybren.local/repos/mpire/test_file.py", line 15, in main
    results = pool.map(square, range(1000), progress_bar=True)
  File "/home/sybren.local/repos/mpire/mpire/pool.py", line 258, in map
    results = self.map_unordered(func, ((args_idx, args) for args_idx, args in enumerate(iterable_of_args)),
  File "/home/sybren.local/repos/mpire/mpire/pool.py", line 314, in map_unordered
    return list(self.imap_unordered(func, iterable_of_args, iterable_len, max_tasks_active, chunk_size,
  File "/home/sybren.local/repos/mpire/mpire/pool.py", line 519, in imap_unordered
    self._handle_exception(progress_bar_handler)
  File "/home/sybren.local/repos/mpire/mpire/pool.py", line 584, in _handle_exception
    raise err("Exception occurred in one of the workers") from err(traceback_str)
ValueError: Exception occurred in one of the workers
```

Is this what you had in mind?

**Comment by remondn-swk** [comment]
@sybrenjansen Actually the first message seems more clear...

So maybe `raise from` is not a good solution.

Actually I opened this issue because in my terminal the error is not clean (`/n` are not interpreted as new line). So I thought it's because of not using `raise from`, but it seems more related to some configuration of my terminal, right ?

Closing the issue as it's not a problem with the library itself.
